sw/device/silicon_creator/lib/sigverify/sphincsplus/hash.h

```diff
@@ -0,0 +1,61 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Derived from code in the SPHINCS+ reference implementation (CC0 license):
+// https://github.com/sphincs/sphincsplus/blob/ed15dd78658f63288c7492c00260d86154b84637/ref/hash.h
+
+#ifndef OPENTITAN_SW_DEVICE_SILICON_CREATOR_LIB_SIGVERIFY_SPHINCSPLUS_HASH_H_
+#define OPENTITAN_SW_DEVICE_SILICON_CREATOR_LIB_SIGVERIFY_SPHINCSPLUS_HASH_H_
+
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/silicon_creator/lib/error.h"
+#include "sw/device/silicon_creator/lib/sigverify/sphincsplus/context.h"
+#include "sw/device/silicon_creator/lib/sigverify/sphincsplus/params.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * Initialize the SPHINCS+ hash function.
+ *
+ * Hook for any precomputation or setup tasks the hash function needs to do
+ * before a SPHINCS+ operation takes place.
+ *
+ * Appears in reference code as `initialize_hash_function`.
+ *
+ * @param ctx Context object.
+ * @return Error code indicating if the operation succeeded.
+ */
+OT_WARN_UNUSED_RESULT
+rom_error_t spx_hash_initialize(spx_ctx_t *ctx);
+
+/**
+ * Hash the input message and derive the leaf index.
+ *
+ * Computes H(R, pk, m), where R is the random number included in the SPHINCS+
+ * signature and H is the underlying hash function (Hmsg in the SPHINCS+
+ * paper). Outputs the message digest and the index of the leaf. The index is
+ * split in the tree index and the leaf index, for convenient copying to an
+ * address.
+ *
+ * @param R Per-signature random number.
+ * @param pk Public key.
+ * @param m Input message.
+ * @param[out] digest Output buffer for message digest.
+ * @param[out] tree Tree index.
+ * @param[out] leaf_idx Leaf index.
+ * @return Error code indicating if the operation succeeded.
+ */
+OT_WARN_UNUSED_RESULT
+rom_error_t spx_hash_message(const unsigned char *R, const unsigned char *pk,
```

Comment:
nit: would it be ok to try to use types similar to what we have defined in the kmac driver interface? for example:

```c
rom_error_t spx_hash_message(const uint8_t *R, const uint8_t *pk,
                             const uint8_t *m, size_t mlen,
                             const spx_ctx_t *ctx, uint8_t *digest,
                             uint64_t *tree, uint32_t *leaf_idx);
```

For the `digest` field it seems the driver is optimized for `uint32_t *`. Would it make sense to follow the same approach here?

Reply:
I've changed the `unsigned char` occurrences now. However, when I looked further into changing `digest` to `uint32_t`, I think it may be better to leave it as `uint8_t` here for a few reasons:
- There's no performance benefit from changing to `uint32_t` since we read the digest into the `uint32_t` array `buf` anyway and then later `memcpy` the result to `digest`.
  - This indirection is necessary anyway because the KMAC driver only has a one-shot `squeeze`, and we need to populate three separate output values from the single `squeeze`. 
- The digest length is not always a multiple of word size; it depends on the parameters, but for the set we're using it's 21 bytes. Using words would waste a few bytes but more importantly cause possible confusion about how long the data actually is.
- The digest later gets passed to FORS and is next meaningfully processed in [this routine](https://github.com/jadephilipoom/opentitan/blob/99fef135a9d3f8039b135c7b69391f29f8948c53/sw/device/silicon_creator/lib/sigverify/sphincsplus/fors.c#L39) as the `m` input. Here it is very much treated as bytes; we'd need to simply cast it back to bytes before using it.

Reply:
SGTM. 

Just as a general rule in this architecture. 32bit is optimized for access, transport and security. 8bit operations incur a penalty and are less secure from a leakage perspective. This is not necessarily relevant within the context of this PR, but bringing it up given that you mentioned performance tradeoffs.